[JSC] Improve loop unrolling heuristics by counting only material nodes

hyjorc1 opened the pull request:
#### d8043dd24f7eddc83c06c6730d025e8ce98684c5
<pre>
[JSC] Improve loop unrolling heuristics by counting only material nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=290513">https://bugs.webkit.org/show_bug.cgi?id=290513</a>
<a href="https://rdar.apple.com/147993710">rdar://147993710</a>

Reviewed by Justin Michaud and Yusuke Suzuki.

This patch refines the loop unrolling heuristic in the DFG by counting only
nodes that would actually emit code (&quot;material&quot; nodes) when estimating the
loop body size. This prevents loops from being prematurely disqualified for
unrolling due to non-material nodes such as Phantom or ExitOK.

Previously, loop body size was estimated using BasicBlock::size(), which
includes all nodes, even those that don&apos;t lower to code. This patch introduces
a new helper `isMaterialNode()` to check whether a node is code-generating.
This logic is aligned with DFGDCEPhase and LowerDFGToB3::compileNode.

We also removed the separate `canCloneLoop()` method and merged cloneability
checks directly into `shouldUnrollLoop()` to avoid double iteration and simplify control flow.

* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::shouldUnrollLoop):
(JSC::DFG::LoopUnrollingPhase::isMaterialNode):
(JSC::DFG::LoopUnrollingPhase::canCloneLoop): Deleted.

Canonical link: <a href="https://commits.webkit.org/292899@main">https://commits.webkit.org/292899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0fa3ace0995439f3a7bd9815bf8c149a1fed202

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7119 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102366 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47808 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25354 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74123 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31341 "Found 1 new test failure: http/tests/cookies/document-cookie-during-iframe-parsing.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100283 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13013 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87978 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54460 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12767 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5855 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47251 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89956 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82812 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104386 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/95902 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24422 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17766 "Found 2 new test failures: imported/w3c/web-platform-tests/workers/abrupt-completion.html ipc/send-gpu-GetShareableBitmap-RemoteRenderingBackend.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83200 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24730 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82579 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20802 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27138 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4807 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17902 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24321 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29474 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119528 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24143 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33556 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27457 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25717 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me and maybe we should ensure that the current threshold is fine or not for this new counting

hyjorc1 committed:
[JSC] Improve loop unrolling heuristics by counting only material nodes
https://bugs.webkit.org/show_bug.cgi?id=290513
rdar://147993710

Reviewed by Justin Michaud and Yusuke Suzuki.

This patch refines the loop unrolling heuristic in the DFG by counting only
nodes that would actually emit code ("material" nodes) when estimating the
loop body size. This prevents loops from being prematurely disqualified for
unrolling due to non-material nodes such as Phantom or ExitOK.

Previously, loop body size was estimated using BasicBlock::size(), which
includes all nodes, even those that don't lower to code. This patch introduces
a new helper `isMaterialNode()` to check whether a node is code-generating.
This logic is aligned with DFGDCEPhase and LowerDFGToB3::compileNode.

We also removed the separate `canCloneLoop()` method and merged cloneability
checks directly into `shouldUnrollLoop()` to avoid double iteration and simplify control flow.

* Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp:
(JSC::DFG::LoopUnrollingPhase::tryUnroll):
(JSC::DFG::LoopUnrollingPhase::identifyInductionVariable):
(JSC::DFG::LoopUnrollingPhase::shouldUnrollLoop):
(JSC::DFG::LoopUnrollingPhase::isMaterialNode):
(JSC::DFG::LoopUnrollingPhase::canCloneLoop): Deleted.

Canonical link: https://commits.webkit.org/292899@main
